src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,200 @@
+        return [lower_bound](const ValueID& value_id) { return value_id < lower_bound; };
+      } else {
+        return [lower_bound, upper_bound](const ValueID& value_id) {
+          return value_id < lower_bound || value_id >= upper_bound;
+        };
+      }
+    case ScanType::OpLessThan:
+      return (lower_bound == INVALID_VALUE_ID) ? always(true) : [lower_bound](const ValueID& value_id) {
+        return value_id < lower_bound;
+      };
+    case ScanType::OpLessThanEquals:
+      return (upper_bound == INVALID_VALUE_ID) ? always(true) : [upper_bound](const ValueID& value_id) {
+        return value_id < upper_bound;
+      };
+    case ScanType::OpGreaterThan:
+      return (upper_bound == INVALID_VALUE_ID) ? always(false) : [upper_bound](const ValueID& value_id) {
+        return value_id >= upper_bound;
+      };
+    case ScanType::OpGreaterThanEquals:
+      return (lower_bound == INVALID_VALUE_ID) ? always(false) : [lower_bound](const ValueID& value_id) {
+        return value_id >= lower_bound;
+      };
+    default:
+      throw std::runtime_error("Unknown ScanType operator.");
+  }
+}
+
+// Scans a segment of values and adds their chunk_offset to the returned container
+// if the passed compare operator returned true.
+// The element_accessor std::function parameter is used to access the actual values of the segment.
+template <typename T>
+std::vector<ChunkOffset> scan_segment(std::function<bool(const T&)> compare,
+                                      std::function<const T(const ChunkOffset&)> element_accessor,
+                                      size_t container_size) {
+  std::vector<ChunkOffset> chunk_offsets;
+
+  for (ChunkOffset current_chunk_offset{0}; current_chunk_offset < container_size; current_chunk_offset++)
+    if (compare(element_accessor(current_chunk_offset))) chunk_offsets.push_back(current_chunk_offset);
+
+  return chunk_offsets;
+}
+
+// Scans the input table and creates an output table containing a Chunk of ReferenceSegments pointing to the values
+// that fulfill the compare operator with the given scan_type and search_value
+template <typename T>
+void scan_table(std::shared_ptr<const Table> input_table, const std::shared_ptr<Table>& output_table,
+                const ColumnID& column_id, ScanType scan_type, T typed_search_value) {
+  for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
+    auto& current_chunk = input_table->get_chunk(chunk_id);
+
+    // no need to scan through the chunk if it's empty
+    if (!current_chunk.size()) continue;
+
+    auto current_segment = current_chunk.get_segment(column_id);
+
+    std::vector<ChunkOffset> referenced_chunk_offsets;
+
+    // Calculate referenced chunk offsets by scanning the segment values
+    // Therefore cast the BaseSegment pointer to the appropriate concrete segment type by try-and-error
+    if (auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(current_segment); value_segment) {
```

Comment:
I don't understand the value_segment behind the ; at the end of the line. I think you could remove it.

Reply:
The first part is called 'init-statement' (since c++17), the second part is the actual condition. It allows to do more complex conditions based the declared variable, than a simple null check.
(https://en.cppreference.com/w/cpp/language/if)

Consider the following example:
```
int a = 5;
        
if (int *b = &a; *b > 3) {
  std::cout << "valid" << std::endl;
}
```

However in this case we only do a  null check, so you are right - it could be omitted.

Reply:
I see. Never heard of that before. Thank you! 😊